### PR TITLE
[MOD 14980] TTL table Rust implementation

### PR DESCRIPTION
**Pull request by allevo**
## Describe the changes in the pull request

Initial porting of TTL Table in Rust.
*Don't*:
- micro benchmarks
- introduce ffi
- replace the C implementation
- remove the C implementation

NB: Sonar fails because it doesn't recognize the Rust workspace, so it expects to have the Cargo.lock committed for each workspace member.

#### Main objects this PR modified
1. TTL Table

#### Mark if applicable

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

#### Release Notes

- [ ] This PR requires release notes
- [x] This PR does not require release notes

If a release note is required (bug fix / new feature / enhancement), describe the **user impact** of this PR in the title.  

---

> [!NOTE]
> **Medium Risk**
> Adds a new Rust TTL table implementation with `unsafe` insertion preconditions and extensive bitmask/expiration logic; while not wired into production paths yet, it expands the workspace/build surface and could introduce correctness or compilation/test issues.
> 
> **Overview**
> Introduces a new Rust workspace crate, `ttl_table`, implementing a per-(document, field) TTL table with lazy-growing direct-modulo buckets, collision chains, and helpers to verify expirations for single fields and 32-bit/128-bit field masks.
> 
> Adds comprehensive unit/integration tests for bucket growth, collision handling, mask iteration, and predicate semantics, updates the Rust workspace (`Cargo.toml`/`Cargo.lock`) to include the new crate, and makes a tiny formatting-only tweak in `tests/pytests/test_expire.py`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ac790de71b14bb404d725af52ceb88b9929575ab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1950 @@
...
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+//! A per-(document-field) time-to-live table
+//!
+//! This crate mirrors the C implementation's data layout and growth strategy:
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Rust code and comments should be self contained and not reference the C implementation which will eventually go away.

Multiple comments should be updated.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,1950 @@
...
+#[derive(Debug, Clone, Copy)]
+pub struct FieldExpiration {
+    /// The field index this expiration applies to.
+    pub index: u16,
+    /// The point in time at which the field expires.
+    pub point: timespec,
+}
+
+#[derive(Debug)]
+struct TimeToLiveEntry {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Let's document this struct and its fields as well.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,1950 @@
...
+}
+
+/// Direct-modulo bucket array with contiguous-vec collision chains.
+///
+/// See the module-level documentation for the rationale. The bucket array
+/// length (`buckets.len()`) plays the role of `cap` in the C version and
+/// always satisfies `buckets.len() <= max_size`.
+#[derive(Debug)]
+pub struct TimeToLiveTable {
+    buckets: Vec<ThinVec<TimeToLiveEntry>>,
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs:78]: Couple of fields are missing doc.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,1950 @@
...
+
+impl TimeToLiveTable {
+    /// Creates an empty table with `max_size` as the fixed modulus for
+    /// the slot formula.
+    ///
+    /// The bucket array starts empty and grows on demand.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `max_size == 0`, which would cause a divide-by-zero in
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Let's use `NonZeroUsize` then?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,1950 @@
...
+//! - Lazy growth: the bucket array starts empty and grows geometrically
+//!   (+1.5×, capped at `1 << 20` and clamped to `max_size`) only as `add`
+//!   demands more slots.
+//! - No shrink-on-delete: empty buckets are released, but the bucket array
+//!   itself keeps its high-water-mark length so churning indexes don't
+//!   thrash on realloc.
+//!
+//! The table holds only field-level (HEXPIRE) expirations; document-level
+//! TTL lives directly on `RSDocumentMetadata::expirationTimeNs` so the
+//! result-processor hot path avoids a lookup here.
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:27]: Write doc comments as if the C version doesn't exist, since it will be deleted once this is ready to go 😁

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,1950 @@
...
+    /// Creates an empty table with `max_size` as the fixed modulus for
+    /// the slot formula.
+    ///
+    /// The bucket array starts empty and grows on demand.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `max_size == 0`, which would cause a divide-by-zero in
+    /// the slot formula.
+    #[must_use]
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: No need for `must_use` here and for `is_empty`.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,1950 @@
...
+    pub fn new(max_size: usize) -> Self {
+        assert!(max_size >= 1, "TTL table max_size must be >= 1");
+
+        // So we can safety convert `self.max_size` to u64
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Can this ever fail?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,1950 @@
...
+/// expiration set" and never expires.
+#[inline]
+const fn did_expire(field: &timespec, now: &timespec) -> bool {
+    if field.tv_sec == 0 && field.tv_nsec == 0 {
+        return false;
+    }
+    !((field.tv_sec > now.tv_sec) || (field.tv_sec == now.tv_sec && field.tv_nsec > now.tv_nsec))
+}
+
+#[cfg(test)]
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:614]: As a general pattern, we prefer to test public APIs via tests in the `tests/` folder. We only use unit tests when we have complex functionality that needs testing but can't be easily exercised from the public API boundary.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,1950 @@
...
+pub struct FieldExpiration {
+    /// The field index this expiration applies to.
+    pub index: u16,
+    /// The point in time at which the field expires.
+    pub point: timespec,
+}
+
+#[derive(Debug)]
+struct TimeToLiveEntry {
+    doc_id: u64,
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: I assume that all these `doc_id` variables/fields should use `t_docId` as type?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,1950 @@
...
+    /// `sorted_by_id` must be non-empty and sorted by `index` in ascending
+    /// order. Ownership transfers to the table.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `sorted_by_id` is empty.
+    /// In debug builds, it panics:
+    /// -  if `doc_id` is already present in the table
+    pub fn add(&mut self, doc_id: u64, sorted_by_id: ThinVec<FieldExpiration>) {
+        assert!(
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Any reason to not use `debug_assert` here as well?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,1950 @@
...
+    /// Inserts a document's per-field expirations.
+    ///
+    /// `sorted_by_id` must be non-empty and sorted by `index` in ascending
+    /// order. Ownership transfers to the table.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `sorted_by_id` is empty.
+    /// In debug builds, it panics:
+    /// -  if `doc_id` is already present in the table
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: And if it's not sorted.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,1950 @@
...
+        self.buckets.len()
+    }
+
+    /// Returns the per-field expiration list stored for `doc_id`, or `None`
+    /// if no entry exists.
+    ///
+    /// The slice aliases storage owned by the table and is invalidated by any
+    /// subsequent [`add`](Self::add) / [`remove`](Self::remove) on this table.
+    #[must_use]
+    pub fn get_field_expirations(&self, doc_id: u64) -> Option<&[FieldExpiration]> {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
    pub fn field_expirations(&self, doc_id: u64) -> Option<&[FieldExpiration]> {
```

Rust getters are usually not prefixed with  `get_`

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,1950 @@
...
+impl TimeToLiveTable {
+    /// Creates an empty table with `max_size` as the fixed modulus for
+    /// the slot formula.
+    ///
+    /// The bucket array starts empty and grows on demand.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `max_size == 0`, which would cause a divide-by-zero in
+    /// the slot formula.
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Then let's take https://doc.rust-lang.org/std/num/type.NonZeroUsize.html as input.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,1950 @@
...
+    #[must_use]
+    pub fn new(max_size: usize) -> Self {
+        assert!(max_size >= 1, "TTL table max_size must be >= 1");
+
+        // So we can safety convert `self.max_size` to u64
+        assert!(
+            u64::try_from(max_size).is_ok(),
+            "max_size is bigger than u64"
+        );
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: What are we guarding against here, exactly? A platform where `usize` is larger than 8 bytes?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,1950 @@
...
+    ) -> bool {
+        let Some(entry) = self.find_entry(doc_id) else {
+            // the document did not have a ttl for itself or its fields
+            // if predicate is FieldExpirationPredicate::Default, at least one field is valid
+            // if predicate is FieldExpirationPredicate::Missing, the field is indeed missing since the document has no expiration for it
+            return true;
+        };
+
+        // This never happen, it is just a guard
+        if entry.field_expirations.is_empty() {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Use a `debug_assert` instead then?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,1950 @@
...
+            // The document has no fields with expiration times, there exists at least one valid field
+            return true;
+        }
+
+        for fe in entry.field_expirations.iter() {
+            if fe.index == field {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Can we use `any()` on the iterator instead?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,1950 @@
...
+    /// `ft_id_to_field_index[bit]` translates a bit position in the mask
+    /// into the `t_fieldIndex` recorded in the table; it must contain at
+    /// least as many entries as the highest set bit of `field_mask`. The
+    /// translation is required to be monotonic, bits scanned low-to-high
+    /// must produce non-decreasing field indices.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `ft_id_to_field_index` is shorter than `highest_set_bit + 1`.
+    #[must_use]
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
```

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,1950 @@
...
+
+    /// Checks the expiration state of a set of fields described by a
+    /// 128-bit `field_mask` (the wide-schema variant).
+    /// See also [`TimeToLiveTable::verify_doc_and_field_mask`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if `ft_id_to_field_index` is shorter than
+    /// `highest_set_bit + 1` of `field_mask`.
+    #[must_use]
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
```

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,1950 @@
...
+    }
+
+    /// Direct-modulo slot formula
+    #[inline]
+    const fn slot(&self, doc_id: u64) -> usize {
+        // All the casts here are safe because we check the conversion in the constructor
+        if doc_id < self.max_size as u64 {
+            doc_id as usize
+        } else {
+            (doc_id % self.max_size as u64) as usize
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:306]: Since the divisor is known and stable, we may even try something like https://crates.io/crates/strength_reduce.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,1950 @@
...
+        .max()
+        .unwrap_or(0);
+    // Hoisted bound. The inner loop derives `bit_index` from
+    // `trailing_zeros` over the same halves used to compute
+    // `highest_bit_plus_one`, but LLVM cannot relate those two
+    // bit-manipulation intrinsics on its own — so the per-iteration
+    // bounds check survives without a hint. This assertion both
+    // gives the caller an actionable error message and underwrites
+    // the `unsafe` proof at the call site of `get_field_index`.
+    assert!(
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs:515]:
```suggestion
    assert_debug!(
```

? We usually avoid assertion in release build for performances.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,1950 @@
...
+    /// Reads `ft_id_to_field_index[bit_index]` while telling LLVM the
+    /// bound holds, so the per-call bounds check on the slice indexing
+    /// is elided.
+    ///
+    /// # Safety
+    ///
+    /// The caller must guarantee `bit_index < ft_id_to_field_index.len()`.
+    /// `assert_unchecked` is a *hint*.
+    #[inline]
+    unsafe fn get_field_index(ft_id_to_field_index: &[u16], bit_index: usize) -> u16 {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
    unsafe fn get_field_index_unchecked(ft_id_to_field_index: &[u16], bit_index: usize) -> u16 {
```

maybe? Same for the next one.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 22]
```diff
@@ -0,0 +1,1950 @@
...
+#[inline]
+const fn did_expire(field: &timespec, now: &timespec) -> bool {
+    if field.tv_sec == 0 && field.tv_nsec == 0 {
+        return false;
+    }
+    !((field.tv_sec > now.tv_sec) || (field.tv_sec == now.tv_sec && field.tv_nsec > now.tv_nsec))
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs:615]: Tests relying only on the public API should be in `tests/integration`.

File: tests/pytests/test_expire.py [hunk 23]
```diff
@@ -590,6 +590,79 @@ def testLastFieldNoExpiration(env):
     env.expect('FT.SEARCH', 'idx', 'hello', 'NOCONTENT').apply(sort_document_names).equal([2, 'doc:1', 'doc:2'])
 
 
+@skip(cluster=True, redis_less_than='7.4')
+def testWideSchemaFieldExpiration(env):
```
Comment (gdesmott) [inline tests/pytests/test_expire.py]: May be worth its own commit/PR so we can merge this already.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 24]
```diff
@@ -0,0 +1,1950 @@
...
+                }
+                current_field_index += 1;
+            }
+            if current_field_index >= field_with_expiration_length {
+                // No more fields with expiration times to check.
+                break 'outer;
+            }
+
+            // SAFETY: the `if … break 'outer` above ensures we only get
+            // here when `current_field_index < field_with_expiration_count`
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
            // here when `current_field_index < field_with_expiration_length.`
```

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,1950 @@
...
+
+        debug_assert!(
+            !self.buckets[slot].iter().any(|e| e.doc_id == doc_id),
+            "duplicate docId in TTL table"
+        );
+
+        self.buckets[slot].push(TimeToLiveEntry {
+            doc_id,
+            field_expirations: sorted_by_id,
+        });
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:146]: Do we care more about the cost of insertion or the cost of retrieval?
If the latter, it may be worthwhile to keep each bucket sorted by doc_id, thus allowing us to binary search rather than linear scanning on the retrieval side.
But the benefits obviously depend on the expected size of these entries.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,1950 @@
...
+    /// because none of the document's fields is expired, so the query trivially passes
+    /// under either predicate.
+    #[must_use]
+    pub fn verify_doc_and_field(
+        &self,
+        doc_id: u64,
+        field: u16,
+        predicate: FieldExpirationPredicate,
+        expiration_point: &timespec,
+    ) -> bool {
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:210]: Naming-wise, having a `verify_...` method that returns a boolean makes it rather unclear what `true`/`false` output means.
I wonder if we should take `predicate` as an input at all. We could simplify this down to an `Option<ExpirationStatus>` and then let the caller decide what they care about.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 27]
```diff
@@ -0,0 +1,1950 @@
...
+    );
+
+    /// Reads `ft_id_to_field_index[bit_index]` while telling LLVM the
+    /// bound holds, so the per-call bounds check on the slice indexing
+    /// is elided.
+    ///
+    /// # Safety
+    ///
+    /// The caller must guarantee `bit_index < ft_id_to_field_index.len()`.
+    /// `assert_unchecked` is a *hint*.
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: We no longer use assert_unchecked, no?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,1950 @@
...
+impl BitMask for u128 {
+    type Halves = [u64; 2];
+
+    #[inline]
+    fn halves(self) -> Self::Halves {
+        [self as u64, (self >> 64) as u64]
+    }
+}
+
+fn verify_mask<M: BitMask>(
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:483]: This function is quite a beast. Any reasonable way to decompose it?

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 26 above)
Comment (LukeMathWalker) [reply src/redisearch_rs/ttl_table/src/lib.rs:210]: If we are pushing the predicate down as an optimization opportunity, then we might want to leverage const generics to dispatch to an optimized version for each of the two possible predicate policies.

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 8 above)
Comment (allevo) [reply src/redisearch_rs/ttl_table/src/lib.rs:614]: I kept here only the ones that require an access to private method like `slot`

File: tests/pytests/test_expire.py (hunk 23 above)
Comment (allevo) [reply tests/pytests/test_expire.py]: Right. https://github.com/RediSearch/RediSearch/pull/9429

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 29]
```diff
@@ -0,0 +1,741 @@
...
+        let bucket = &mut self.buckets[slot];
+        if let Some(pos) = bucket.iter().position(|e| e.doc_id == doc_id) {
+            bucket.swap_remove(pos);
+            self.count -= 1;
+        }
+    }
+
+    /// Return the number of buckets currently allocated
+    #[doc(hidden)]
+    pub const fn allocated_buckets(&self) -> usize {
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
    pub const fn n_allocated_buckets(&self) -> usize {
```

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 30]
```diff
@@ -0,0 +1,741 @@
...
+//! - No shrink-on-delete: empty buckets are released, but the bucket array
+//!   itself keeps its high-water-mark length so churning indexes don't
+//!   thrash on realloc.
+//!
+//! The table holds only field-level (HEXPIRE) expirations; document-level
+//! TTL lives directly on `RSDocumentMetadata::expirationTimeNs` so the
+//! result-processor hot path avoids a lookup here.
+
+#[doc(hidden)]
+pub mod test_utils;
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: Proper feature flag please.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 31]
```diff
@@ -0,0 +1,741 @@
...
+        let slot = self.slot(doc_id);
+        if slot >= self.buckets.len() {
+            return;
+        }
+        let bucket = &mut self.buckets[slot];
+        if let Some(pos) = bucket.iter().position(|e| e.doc_id == doc_id) {
+            bucket.swap_remove(pos);
+            self.count -= 1;
+        }
+    }
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs:168]: Better to return `Option<..>` with the removed entry, so that a caller can actually determine whether or not something was removed. We get it for free in `bucket.swap_remove(pos)`.

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 28 above)
Comment (allevo) [reply src/redisearch_rs/ttl_table/src/lib.rs:483]: I use an external iterator without performance penalty to make the loop easier to read.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 32]
```diff
@@ -0,0 +1,832 @@
...
+        if !expired && predicate == FieldExpirationPredicate::Default {
+            return true;
+        }
+        if expired && predicate == FieldExpirationPredicate::Missing {
+            return true;
+        }
+        predicate_misses += 1;
+    }
+
+
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table/src/lib.rs]: redundant \n

File: src/redisearch_rs/ttl_table_bencher/src/lib.rs [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+//! Shared helpers for the [`ttl_table`] Criterion benches.
+//!
+//! The TTL table is sized by the host `DocTable`'s `maxSize`
+//! (`DEFAULT_DOC_TABLE_SIZE = 1_000_000` in `src/config.h`) and is populated
+//! sparsely with only those documents that have HEXPIRE'd fields. DocIds are
+//! allocated monotonically by the indexer, so the production hot path inserts
+//! and looks up roughly sequential IDs against a direct-modulo bucket array.
+//! These helpers default to that pattern.
+
+use std::num::NonZeroUsize;
```
Comment (gdesmott) [inline src/redisearch_rs/ttl_table_bencher/src/lib.rs]: let's move the benchs to another PR. This one is already big enough.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 34]
```diff
@@ -0,0 +1,741 @@
...
+    ///
+    /// Panics if `sorted_by_id` is empty.
+    /// In debug builds, it panics:
+    /// - if `doc_id` is already present in the table
+    /// - if `sorted_by_id` is not sorted by index
+    pub fn add(&mut self, doc_id: t_docId, sorted_by_id: ThinVec<FieldExpiration>) {
+        debug_assert!(
+            !sorted_by_id.is_empty(),
+            "TTL table add requires at least one field expiration"
+        );
```
[the remark this answers is not in the document]
Comment (gdesmott) [reply src/redisearch_rs/ttl_table/src/lib.rs:129]: Sounds legit.

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 26 above)
Comment (gdesmott) [reply src/redisearch_rs/ttl_table/src/lib.rs:210]: I agree something like `is_expired()` would be much clearer.

Or `expiration_status()` with the API change you suggest. I link the idea, but I think I'd go with an enum covering all cases rather than wrapping it in an `Option`.

File: src/redisearch_rs/ttl_table_bencher/src/lib.rs (hunk 33 above)
Comment (allevo) [reply src/redisearch_rs/ttl_table_bencher/src/lib.rs]: Removed from this PR

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 35]
```diff
@@ -0,0 +1,746 @@
...
+    // `trailing_zeros` over the same halves used to compute
+    // `highest_bit_plus_one`, but LLVM cannot relate those two
+    // bit-manipulation intrinsics on its own — so the per-iteration
+    // bounds check survives without a hint. This assertion both
+    // gives the caller an actionable error message and underwrites
+    // the `unsafe` proof at the call site of `get_field_index`.
+    debug_assert!(
+        ft_id_to_field_index.len() >= highest_bit_plus_one,
+        "ft_id_to_field_index must cover the highest set bit of the mask"
+    );
```
[the remark this answers is not in the document]
Comment (LukeMathWalker) [reply src/redisearch_rs/ttl_table/src/lib.rs:518]: Looking through it, this is indeed true, no? @allevo

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 36]
```diff
@@ -0,0 +1,830 @@
...
+    let field_expirations: &[FieldExpiration] = &entry.field_expirations;
+    debug_assert!(
+        !entry.field_expirations.is_empty(),
+        "field_expirations is guaranteed to not be empty"
+    );
+    let field_with_expiration_length = field_expirations.len();
+
+    let halves = mask.halves();
+    let halves = halves.as_ref();
+    let field_count: usize = halves.iter().map(|h| h.count_ones() as usize).sum();
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: We have `count_ones` as a straight method on `u128`, so let's lift it to the `BitMask` trait and invoke it directly on `mask`.

Comment (LukeMathWalker) [reply src/redisearch_rs/ttl_table/src/lib.rs]: Also, this is only needed if the predicate is set to `Default`.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 37]
```diff
@@ -0,0 +1,830 @@
...
+            bit_index < ft_id_to_field_index.len(),
+            "Safety violation: try to access to an out-of-bounds index, {bit_index}. Length {}",
+            ft_id_to_field_index.len(),
+        );
+        // SAFETY: Function safety guarantees
+        unsafe { *ft_id_to_field_index.get_unchecked(bit_index) }
+    }
+
+    /// Reads `&field_expirations[field_index]` while telling LLVM the
+    /// bound holds, so the per-call bounds check is elided.
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: This is not true. We simply read without bounds checks.

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 38]
```diff
@@ -0,0 +1,830 @@
...
+
+    let mut predicate_misses: usize = 0;
+    let mut current_field_index: usize = 0;
+
+    // Outer loop walks halves. This is known at compile time:
+    // u32 -> `[u64; 1]`, u128 -> `[u64; 2]`.
+    // NB: in case of u128, the order  (lower bits, higher bits) is important.
+    // In fact, we can keep track of `current_field_index` and perform a fast check.
+    // NB2: the outer loop is fully unrolled by LLVM.
+    'outer: for (half_idx, &half_init) in halves.iter().enumerate() {
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]: We have the `BitU128Iter` version, but we don't actually use it here?

File: src/redisearch_rs/ttl_table/src/lib.rs [hunk 39]
```diff
@@ -0,0 +1,830 @@
...
+        );
+        assert!(t.verify_doc_and_field(2, FIELD_INDEX_1, FieldExpirationPredicate::Default, &NOW,));
+
+        assert_eq!(t.slot(0), t.slot(1));
+        assert_eq!(t.slot(0), t.slot(2));
+    }
+
+    #[test]
+    #[should_panic(expected = "field_expirations is guaranteed to not be empty")]
+    fn verify_mask_returns_true_when_entry_has_no_field_expirations() {
```
Comment (LukeMathWalker) [inline src/redisearch_rs/ttl_table/src/lib.rs]:
```suggestion
    fn verify_mask_panics_when_entry_has_no_field_expirations() {
```

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 19 above)
Comment (allevo) [reply src/redisearch_rs/ttl_table/src/lib.rs:306]: I'll do it in another pr

File: src/redisearch_rs/ttl_table/src/lib.rs (hunk 25 above)
Comment (allevo) [reply src/redisearch_rs/ttl_table/src/lib.rs:146]: I'll do it in another PR

Comment (ofiryanai) [reply src/redisearch_rs/ttl_table/src/lib.rs:146]: expected size of these entries is very tiny - 1~4 I'd say. doc_id is monotonic increasing, and the "hash" is simply modulu of the index size. no need to sort IMO

**Review by ofiryanai — DISMISSED** [review]
LGTM, please make sure the CI pass so you can merge. especially liked the `ThinVec` (looks like clean win in terms of size if most are empty), the `NonZeroUsize` and  the BitMask tricks!

**Review by gdesmott — APPROVED** [review]
